File: Data Structures & Algorithms/coin-change-ii/submission-7.cpp
class Solution {
public:
    vector<vector<int>> memo;
    int dfs(int i, int a , vector<int>& coins){
        if(a==0) return 1;
        if(i >= coins.size()) return 0;
        int res=0;
        if(memo[i][a]!=-1) return memo[i][a];
        if(a >= coins[i]){
            res=dfs(i+1,a,coins);
            res+=dfs(i,a-coins[i],coins);
        }
        memo[i][a]=res;
        return res;
    }
    int change(int amount, vector<int>& coins) {
        sort(coins.begin(),coins.end());
        memo.assign(coins.size()+1, vector<int>(amount+1, -1));
        return dfs(0,amount , coins);
    }
};

File: Data Structures & Algorithms/three-integer-sum/submission-7.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        if nums[0] > 0 or nums==[] or ( nums[0]==0 and nums[2]>0 ) :
            return []
        else :
            R=[]
            S=[]
            for i in range(len(nums)) :
                k,j=i+1 ,len(nums)-1
                while k < j :
                    if nums[i]+nums[j]+nums[k] == 0 :
                        if{nums[i],nums[j],nums[k]} not in S :
                            R.append([nums[i],nums[j],nums[k]])
                            S.append({nums[i],nums[j],nums[k]})
                        j-=1
                    elif nums[i]+nums[j]+nums[k] > 0 :
                        j-=1
                    elif nums[i]+nums[j]+nums[k] < 0 :
                        k+=1
            return R